### PR TITLE
Hide paged items in Pagination component on narrow viewports

### PR DESCRIPTION
**Pull request by rezrah**
## Summary

Resolves https://github.com/github/primer/issues/3744

Numbered pagination items are now hidden on narrow viewports by default. This prevents horizontal overflow / scrolling issues when a large number of items are present. 

While the API already supports `showPages`, this enforces that behavior by default while retaining the option to customize it.

🔗  [Storybook preview](https://primer-ce1b22441a-26139705.drafts.github.io/brand/storybook/?path=/story/components-pagination-features--narrow-page-numbers-hidden-by-default&globals=viewport:iphonexr)

## List of notable changes:

- `showPages` now undefined by default (prev `true`)
- `showPages` is `true` if viewport is > `isMedium`
- `showPages` continues to be customizable and forwarded by end-users (backwards compatible)

## What should reviewers focus on?

- Are there any edge cases that haven't been handled?

## Steps to test:

1. [Use the dedicated story](https://primer-ce1b22441a-26139705.drafts.github.io/brand/storybook/?path=/story/components-pagination-features--narrow-page-numbers-hidden-by-default&globals=viewport:iphonexr) to see new default behaviour
1. Scroll through all other stories and check the behavior matches the expectations (hidden on mobile, still visible on medium+)

## Supporting resources (related issues, external links, etc):

- https://github.com/github/primer/issues/3744

## Contributor checklist:

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] New visual snapshots have been generated / updated for any UI changes
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Screenshots:

> Please try to provide before and after screenshots or videos

<table>
<tr>
<th> Before</th> <th> After </th>
</tr>
<tr>
<td valign="top">

image

 </td>
<td valign="top">

Screenshot 2024-08-23 at 11 05 31

</td>
</tr>
</table>

### CODE REVIEW

File: packages/react/src/BreakoutBanner/BreakoutBanner.test.tsx [hunk 1]
```diff
@@ -76,6 +76,7 @@ describe('BreakoutBanner', () => {
   it('renders primary link with arrow by default', () => {
     const {getAllByRole} = render(
       <BreakoutBanner>
+        <BreakoutBanner.Heading>Required heading</BreakoutBanner.Heading>
```
Comment (rezrah) [inline packages/react/src/BreakoutBanner/BreakoutBanner.test.tsx:79]: FYI: Unrelated to the pagination change. This fixes some noisy test logs.

**Review by simmonsjenna — APPROVED** [review]
Looks good! 👍
